Add eSealing tutorial (#727)
Concordium/concordium.github.io: source/mainnet/smart-contracts/tutorials/eSealing/eSealing_dapp.rst

.. include:: ../../../variables.rst
.. _eSealing_dapp:

=============
eSealing dApp
=============

You can explore the `hosted eSealing service <https://esealing.testnet.concordium.com>`_ or

> start your own front-end by following the instructions in the READ.me file of this

Reviewer: ```suggestion
start your own front-end by following the instructions in the README.md file of this
```

`repository <https://github.com/Concordium/concordium-browser-wallet/tree/main/examples/eSealing>`_ to get an overview of the dApp flow.

.. note::

   Comprehensive instructions on how to set up the |bw|, create an account in the |bw|,

>    get some testnet CCD and run a local front-end can be found in :ref:`wCCD front-end-set-up section <wCCD-front-end-set-up>`

Reviewer: ```suggestion
   get some testnet CCD and run a local front-end can be found in :ref:`wCCD front-end-set-up section <wCCD-front-end-set-up>`.
```

The front end supports the following two flows with the |bw| (or |mw-gen2| that uses WalletConnect):

- Compute the file hash of a selected file => register its file hash in the smart contract
- Compute the file hash of a selected file => retrieve the timestamp and witness (sealer account) that registered the file hash

.. note::

>    Select the ``registration tab`` to register the file hash and the ``display tab`` to look up the

Reviewer: We use bold text in a few places to refer to things that users should click on. Perhaps we should also do that here. Do we have any guidelines for that @dg-concordium ?

Author: @dg-concordium Any guidelines/preferences please?

   timestamp and witness (sealer account) of an already registered file.

.. note::

>    A file hash is an algorithmic way to assign a unique string to a file using a cryptographic hash function (e.g., sha256) based on the file content.

Reviewer: We could link to https://en.wikipedia.org/wiki/SHA-2 here.

   Changing the file's content would result in a completely new file hash. The eSealing smart contract expects a file hash of length 256-bits as an

>    input parameter because of the used `HashSha2256 <https://docs.rs/concordium-std/latest/concordium_std/struct.HashSha2256.html>`_ type.

Reviewer: This link seems to be at the wrong level of abstraction for this tutorial. I would suggest just referring to the wiki link I suggest above for the number of bits.

Author: I feel that because `HashSha2256` is a custom type, this link is useful (e.g. the first sentence "Sha2 digest with 256 bits (32 bytes)." was very helpful when I first checked our documentation).

Register a file
===============

Select the ``registration tab`` to register the file hash.

.. image:: ./images/registrationTab.gif
   :alt: StreamPlayer
   :align: center

> Registration Tab

Reviewer: Have you tried using a [figure](https://docutils.sourceforge.io/docs/ref/rst/directives.html#figure) instead of a raw `image`? The figure allows you to add a caption, so it seems like the right thing to use here.

Reviewer: @dg-concordium suggests making the text refer to the image `in the image below you can see ...` instead of using captions. So perhaps that is the way to go.

View timestamp and witness
==========================

Select the ``display tab`` to look up the timestamp and witness (sealer account) of an already registered file.

.. image:: ./images/displayTab.gif
   :alt: StreamPlayer
   :align: center

Display Tab

Use Concordium Client
=====================

The front end is connected to a deployed eSealing smart contract on the Concordium testnet
with `this source code <https://github.com/Concordium/concordium-rust-smart-contracts/tree/main/examples/eSealing>`_.

You can interact directly with the smart contract using ``concordium-client`` and your local node
that is running at port 10001 to register a file hash as follows:

.. code-block:: console

   $concordium-client contract update 2481 --entrypoint registerFile --sender <YourAccount> --energy 30000 --parameter-json fileHash.json --grpc-port 10001

Create a ``fileHash.json`` file with a file hash as content similar to:

.. code-block:: json

>    "14fe0aed941aa0a0be1118d7b7dd70bfca475310c531f1b5a179b336c075db65"

Reviewer: We could tell them how to do this, at least on UNIX, with `sha256sum`.

.. note::

   Comprehensive instructions on how to set up a local Concordium testnet node on port 10001 (or alternatively link to a remote node via grpc),
   and download ``concordium-client`` can be found in :ref:`part 3<piggy-bank-preparing>` of the piggy bank tutorial.

> You can view the ``timestamp`` and ``witness (sealer account)`` of an already timestamped file as follows:

Reviewer: ```suggestion:
You can view the ``timestamp`` and ``witness``(sealer account) of an already timestamped file as follows:
```

.. code-block:: console

>     $concordium-client contract invoke 2481 --entrypoint getFile --parameter-json fileHash.json --grpc-port 10001

Reviewer: I would show an example output here.

What is the cost of using your account to seal a file?
======================================================

The cost is currently approximately 4-5 cents (Euro) to register a file hash in the smart contract on mainnet.

> Displaying the timestamp and witness (sealer_account) of a file hash is free of charge.

Reviewer: Perhaps adding something like:

.. because it doesn't make a transaction, it just looks up the values on the chain.
